src/main/flight/autopilot.c

```diff
@@ -22,30 +22,94 @@
 
 #include "platform.h"
 #include "build/debug.h"
+#include "common/filter.h"
 #include "common/maths.h"
+#include "fc/rc.h"
+#include "fc/runtime_config.h"
 
 #include "flight/imu.h"
-#include "flight/pid.h"
 #include "flight/position.h"
 #include "rx/rx.h"
+#include "sensors/gyro.h"
+#include "sensors/compass.h"
 
 #include "autopilot.h"
 
 #define ALTITUDE_P_SCALE  0.01f
 #define ALTITUDE_I_SCALE  0.003f
 #define ALTITUDE_D_SCALE  0.01f
 #define ALTITUDE_F_SCALE  0.01f
+#define POSITION_P_SCALE  0.001f
+#define POSITION_I_SCALE  0.0004f
+#define POSITION_D_SCALE  0.003f
+#define POSITION_J_SCALE  0.0008f
 
 static pidCoefficient_t altitudePidCoeffs;
+static pidCoefficient_t positionPidCoeffs;
+
 static float altitudeI = 0.0f;
 static float throttleOut = 0.0f;
 
+typedef struct {
+    uint32_t distanceCm;
+    uint32_t previousDistanceCm;
+    float previousDistancePitch;
+    float previousDistanceRoll;
+    float previousVelocityPitch;
+    float previousVelocityRoll;
+    float previousHeading;
+    float pitchI;
+    float rollI;
+    bool justStarted;
+} posHoldState;
+
+static posHoldState posHold = {
+    .distanceCm = 0,
+    .previousDistanceCm = 0,
+    .previousDistancePitch = 0.0f,
+    .previousDistanceRoll = 0.0f,
+    .previousVelocityPitch = 0.0f,
+    .previousVelocityRoll = 0.0f,
+    .previousHeading = 0.0f,
+    .pitchI = 0.0f,
+    .rollI = 0.0f,
+    .justStarted = true,
+};
```

Comment:
These values only have to do with piloting the quad while in position hold.
pos_hold.c just tells autopilot what the initial target location is.
These values are used only in autopilot.c to stay at that location.

If the location was to change, then the PIDs would fly to the new position.  It would require an additional command but the controller could be easily adapted to that task, and would use the same kind of PIDs and controls.  No point duplicating them just because with this function we want to stay still.

Reply:
Where these things go can be debated endlessly.

One function of an autopilot is to 'stay put', which is the 'positionControl` function here.  Another could be 'control velocity (disregard direction)` which we do in GPS rescue a lot, but is not yet in autopilot.c.  Another could be 'fly to location'.  Another could be 'altitudeControl', which we currentlyuse both in GPS Rescue and Altitude Hold.

It seems to me that these functions all belong in one place; currently they are in the 'autopilot' files.

And then, the settings for these functions should be in structs that have names related to the function.

In that sense, it seems sensible that pos_hold.c requests the positionHold function in autopilot, when it needs it.  If we had a 'fly to location' mode, and once it got there it was supposed to hold position, then it makes sense that the hold position code exists in autopilot also.

That's my thinking, anyway, and that's why these structs have posHold names at this point.

When we resolve the conceptual model we can re-name things to match the model.  Let's agree on the model first.

Reply:
First priority is performance, safety and basic code optimisation.

Reply:
The general function of autopilot is supposed to be piloting the drone by itself in general (Abstraction of all autonomous features). Staying put is a "subset" of autopilot. Ideally, this means pos hold is just a special case of the autopilot, in which pos hold tells the autopilot that it wishes to stay put (by holding the position setpoint constant).

I know it is much work to change the underlying structure, so I agree to leaving the PR in its current "shape". Once this is merged, I am going to try to refactor what we have as good as possible. I think that will save all of us a lot of time and effort and is the most efficient and easy way forward.